### ISSUE TITLE
Build, Link errors on Windows and Linux

### ISSUE THREAD

**Issue by rachelll4** [zeromq/cppzmq#368]
Hopefully this is a silly question. 

I built cppzmq on Windows using the libzmq download, then put the github (here) zmq.hpp and zmq.h files in my project directories, and made sure that Visual Studio knew to check that directory for dependencies [Properties-->C/C++ Properties-->Additional Include Directories]. (I'm writing a simple test to get a C++ project (cppzmq) talking to a C# project (just C# bindings of zmq) as a foundation for a larger test.) 

Here's the complete C++ code:
`#include "zmq.hpp"
#include <iostream>
#include <fstream>
using namespace std;

int main()
{
        std::cout << "Hello World!\n";

	//prove opened from c#
	ofstream debug;
	debug.open("run.txt");
	debug << "i am opened"<<endl;
	
	//connect to 5556
	zmq::context_t cppContext(1);
	zmq::socket_t cppSocket(cppContext, ZMQ_PAIR);
	cppSocket.connect("tcp://127.0.0.1:5556");

	//receive init message
	zmq::message_t msg;
	cppSocket.recv(msg);
	debug << msg.data();
	debug.close();

	//reply with yay
	string toSend = "gotcha. back at ya";
	zmq::message_t sent = zmq::message_t(toSend);
	cppSocket.send(sent,zmq::send_flags::none);

	//close
	cppContext.close();
	return 0;
}`

C++ strikes again: I got 15 of these LNK2019 errors: 

> unresolved external symbol __imp__zmq_msg_recv referenced in function "public: class zmq::detail::trivial_optional<unsigned int> __thiscall zmq::detail::socket_base::recv(class zmq::message_t &,enum zmq::recv_flags)" (?recv@socket_base@detail@zmq@@QAE?AV?$trivial_optional@I@23@AAVmessage_t@3@W4recv_flags@3@@Z)

Because this test is ultimately going to have to work on Linux (two steps down the road), I moved to Linux Mint, which is basically Ubuntu 18.04.

I started with the libzmq and cppzmq github downloads outlined in your readme file, but when I saw in various other github issues that we're to use the package instead, i did `sudo apt-get install libzmq3-dev` and deleted the previous file folders I had cloned. I put this code into Code::Blocks, put zmq.hpp and zmq.h in both the dev folder and the bin/Debug folder for luck, and tried to build--

I got the same LNK2019 errors. Since I switched IDE and OS, I sure it's something I'm doing wrong with the header file or the library/package installations. Any pointer/suggestions/instructions?

**Comment by sigiesec** [comment]
Either you do not link against libzmq at all, or you are using a static libzmq. To use that on Windows, you need to define ZMQ_STATIC.

**Comment by rachelll4** [comment]
Would that definition be within my application, say, in the header file or the main file? Would I simply define it as true? How would I do that in Linux?

**Comment by sigiesec** [comment]
Usually, I wouldn't define this in code, but somewhere in the project files. If you manually create the VS solution, then in the VS project definition. If you generate that, e.g. with cmake, then in the CMakeLists.txt. On Linux and other non-Windows systems, the ZMQ_STATIC define is not necessary AFAIK, since linking works differently there.

Note that this is more a libzmq question than a cppzmq question.

**Comment by rachelll4** [comment]
I defined ZMQ_STATIC in project properties [C/C++-->Preprocessor-->Preprocessor Definitions], but while that helped, it didn't solve it. Posting solution here for completion's sake.

I found [this tutorial](https://joshuaburkholder.com/wordpress/2018/05/25/build-and-static-link-zeromq-on-windows/) which pointed out that not only did I have to add an additional include directory (libzmq directory), but I had to add additional dependency (with entire filepath/name): libzmq-v141-mt-4_3_2.lib, as well as the above preprocessor definition. I didn't make using cmake as the article writer did, and changing the runtime library mode only caused more issues. 

This is with Visual Studio 2019 on Windows. I assume my Linux error is similar in nature. Thanks for the help.

**Comment by sigiesec** [comment]
That's what I meant by "you do not link against libzmq at all". Sorry for not being more explicit about that. Glad to hear you solved your issue!

**Comment by rachelll4** [comment]
Where are/what are the libraries (static or otherwise) that I would need to include in Linux for cppzmq? My project doesn't have a CMakeLists file. I tried Project Properties-->Linking files I found in /usr/local/lib (libzmq.so, libzmq.so.5, libzmq.a, libzmq.so.5.2.3) as well as doing a #include "zmq.hpp", but I feel as though I'm once again not linking the correct file.

errors: It's saying various argument quantities are incorrect and there are no matching functions to call (x2) and zmq::send_flags has not been declared, for example. I assume these are because the library was incorrectly linked.

Many thanks!
